Bottom: 32c2df5f18d5e8672b01c37f8fd7f9854e5906bb
Top:    c881c2618ad83e3c43a73e6f27af18b7e1124f98
Author: Jason Wang <jasowang@redhat.com>
Date:   2018-05-17 17:43:31 +0800

virtio-net: reset num_buf to 1 after linearizing packet

If we successfully linearize the packets, num_buf were set to zero
which was wrong since we now have only 1 buffer to be used for e.g in
the error path of receive_mergeable(). Zero num_buf will lead the code
try to pop the buffers of next packet and drop it. Fixing this by set
num_buf to 1 if we successfully linearize the packet.

Fixes: 4941d472bf95 ("virtio-net: do not reset during XDP set")
Signed-off-by: Jason Wang <jasowang@redhat.com>


---

diff --git a/drivers/net/virtio_net.c b/drivers/net/virtio_net.c
index 93ec469..d9637ec 100644
--- a/drivers/net/virtio_net.c
+++ b/drivers/net/virtio_net.c
@@ -536,6 +536,8 @@ static struct page *xdp_linearize_page(struct receive_queue *rq,
 		put_page(p);
 	}
 
+	/* receive_mergeable() assumes there's at least one buffer */
+	*num_buf = 1;
 	/* Headroom does not contribute to packet length */
 	*len = page_off - VIRTIO_XDP_HEADROOM;
 	return page;
